lib/cereal/cereal.go

```diff
@@ -619,37 +616,43 @@ LOOP:
 		}
 		logctx.Debugf("Dequeued task %s", t.Name)
 
-		var runCtx context.Context
+		runCtx := taskCompleter.Context()
 		var cancel context.CancelFunc
 		if timeout > 0 {
-			runCtx, cancel = context.WithTimeout(ctx, timeout)
+			runCtx, cancel = context.WithTimeout(runCtx, timeout)
 		} else {
-			runCtx, cancel = context.WithCancel(ctx)
+			runCtx, cancel = context.WithCancel(runCtx)
 		}
+		runTask(runCtx, logctx, exec, t, taskCompleter) // nolint: errcheck
+		cancel()
+	}
+	m.wg.Done()
+}
 
-		result, err := exec.Run(runCtx, &task{backendTask: t})
+func runTask(ctx context.Context, logctx logrus.FieldLogger, exec TaskExecutor, t *backend.Task, taskCompleter backend.TaskCompleter) error {
+	result, err := exec.Run(ctx, &task{backendTask: t})
+	if err != nil {
+		err := taskCompleter.Fail(err.Error())
 		if err != nil {
-			err := taskCompleter.Fail(err.Error())
-			if err != nil {
-				logctx.WithError(err).Error("failed to mark task as failed")
-			}
-		} else {
-			jsonResults, err := jsonify(result)
-			if err != nil {
-				logctx.WithError(err).Error("could not convert returned results to JSON")
-				if err := taskCompleter.Fail(err.Error()); err != nil {
-					logrus.WithError(err).Error("Failed to fail task completer")
-				}
-			}
-			err = taskCompleter.Succeed(jsonResults)
-			if err != nil {
-				logrus.WithError(err).Error("failed to mark task as successful")
+			logctx.WithError(err).Error("failed to mark task as failed")
+			return err
+		}
+	} else {
```

Comment:
Do we need this `else`?

Reply:
i think so. the previous block doesn't return in all cases